### PR TITLE
Add EventsQueue class with 3 push strategies

### PR DESCRIPTION
**Pull request by mauropasse**
Usage example:
```
auto options = rclcpp::ExecutorOptions();
options.queue_options.queue_strategy = rclcpp::QueueStrategy::LIMITED_EVENTS_WITH_TIME_ORDERING;
auto executor1 = std::make_shared<rclcpp::executors::EventsExecutor>(options);

options.queue_options.queue_strategy = rclcpp::QueueStrategy::BOUNDED;
options.queue_options.max_events = 100;
auto executor2 = std::make_shared<rclcpp::executors::EventsExecutor>(options);

// Use default strategy: QueueStrategy::CPU_PERFORMANCE, no need to pass options
auto executor3 = std::make_shared<rclcpp::executors::EventsExecutor>();
```
I ended up creating 3 policies:
`CPU_PERFORMANCE` the default policy, it should peform as good as without this PR (to confirm soon)
`LIMITED_EVENTS_WITH_TIME_ORDERING`  checks before pushing is there are more events than QoS->depth and reorders
`BOUNDED` the queue is hard bounded to a limit. In case of more events, a prune of the queue happens based on QoS.

TODO:

- Run some tests on the single core platform to create some graphs with performances comparison between the different modes.
- Implement for `clients`, `services` and `events`

### CODE REVIEW

**Comment by alsora** [comment]
Some early comments:
 - The events executor should not have any knowledge of which types of queues are available. The current implementation prevents users to define their own policy without modifying the rclcpp code, which is not good.
 A possible solution would be to have a base class `EventsQueue` and then let the users define their own derivate class from that.
 This can be passed to the events executor then.
 
 - It looks to me that a lot of functionalities are being moved from the executor to the queue, is that really needed? It makes the code hard to follow and basically requires the queue to know about which spin implementations are available.
 The queue should only expose very basic API such as push and pop

File: rclcpp/include/rclcpp/executor_options.hpp [hunk 1]
```diff
@@ -24,18 +24,37 @@
...
+};
+
 /// Options to be passed to the executor constructor.
 struct ExecutorOptions
 {
   ExecutorOptions()
   : memory_strategy(rclcpp::memory_strategies::create_default_strategy()),
     context(rclcpp::contexts::get_global_default_context()),
-    max_conditions(0)
+    max_conditions(0), queue_options(QueueOptions())
```
Comment (alsora) [inline rclcpp/include/rclcpp/executor_options.hpp:51]: The `ExecutorOption` structure should only contain stuff that is used by all executors

File: rclcpp/include/rclcpp/executor_options.hpp [hunk 2]
```diff
@@ -24,18 +24,37 @@
...
+  CPU_PERFORMANCE,
+  LIMITED_EVENTS_WITH_TIME_ORDERING,
+  BOUNDED
+};
+
+struct QueueOptions
+{
+  QueueOptions()
+  : queue_strategy(QueueStrategy::CPU_PERFORMANCE),
+    max_events(1000)
```
Comment (alsora) [inline rclcpp/include/rclcpp/executor_options.hpp:38]: Following my comments above, `max_events` should be an internal detail of a bounded queue class

File: rclcpp/include/rclcpp/executors/events_executor.hpp [hunk 3]
```diff
@@ -174,7 +173,7 @@ class EventsExecutor : public rclcpp::Executor
 private:
   RCLCPP_DISABLE_COPY(EventsExecutor)
 
-  using EventQueue = std::queue<rmw_listener_event_t>;
+  using EventQueue = std::deque<rmw_listener_event_t>;
```
Comment (alsora) [inline rclcpp/include/rclcpp/executors/events_executor.hpp:176]: this looks not needed anymore now (and also you can remove the queue/deque include)

File: rclcpp/include/rclcpp/executors/events_executor.hpp [hunk 4]
```diff
@@ -186,14 +185,8 @@ class EventsExecutor : public rclcpp::Executor
...
-    // Event queue mutex scope
-    {
-      std::unique_lock<std::mutex> lock(this_executor->push_mutex_);
-
-      this_executor->event_queue_.push(event);
-    }
-    // Notify that the event queue has some events in it.
-    this_executor->event_queue_cv_.notify_one();
+    // Push event and notify the queue it has some events
+    this_executor->events_queue_->push_and_notify(event);
```
Comment (alsora) [inline rclcpp/include/rclcpp/executors/events_executor.hpp:189]: This should be just a simple push, then the notify should still be done here

File: rclcpp/include/rclcpp/executors/events_executor_entities_collector.hpp [hunk 5]
```diff
@@ -184,6 +184,39 @@ class EventsExecutorEntitiesCollector final
...
+
+  ///
+  /**
+   * Gets the QoS of the entities collector.
+   * This is useful for the events executor, when it has to
+   * decide if keep pushing events from this waitable into the qeueue
+   */
+  RCLCPP_PUBLIC
+  rmw_qos_profile_t
+  get_actual_qos() const
```
Comment (alsora) [inline rclcpp/include/rclcpp/executors/events_executor_entities_collector.hpp:213]: I'm not sure I understand what this is.
Why a waitable has a QOS?

File: rclcpp/src/rclcpp/executors/events_queue.cpp [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+  }
+}
+
+EventsQueue::~EventsQueue()
+{
+  subscription_events_in_queue_map_.clear();
+  waitable_events_in_queue_map_.clear();
+}
+
+void EventsQueue::swap(EventQueue & event_queue)
```
Comment (alsora) [inline rclcpp/src/rclcpp/executors/events_queue.cpp:61]: We don't need to expose that we are doing a swap. That's just an optimization.
This API should have a signature like:
```
std::queue<EventType> EventsQueue::get_all_events()
```

File: rclcpp/src/rclcpp/executors/events_queue.cpp [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+{
+  std::unique_lock<std::mutex> push_lock(push_mutex_);
+  std::swap(event_queue, event_queue_);
+  clear_stats_implem_();
+}
+
+void EventsQueue::wait_for_event(std::chrono::nanoseconds timeout)
```
Comment (alsora) [inline rclcpp/src/rclcpp/executors/events_queue.cpp:68]: `wait_for_event` and `wait_for_event_and_swap` seem not needed to me.
The executor should manage the wait, then call swap (or its renamed version as I suggest above).
It's also ok to have this class without any mutex and let the executor manage that

File: rclcpp/src/rclcpp/executors/events_queue.cpp [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+  std::unique_lock<std::mutex> push_lock(push_mutex_);
+  // We wait here until something has been pushed to the queue
+  // or if the timeout expired
+  event_queue_cv_.wait_for(push_lock, timeout, has_event_predicate);
+  // Time to swap queues as the wait is over
+  std::swap(event_queue, event_queue_);
+  clear_stats_implem_();
+}
+
+bool EventsQueue::wait_and_get_first_event(
```
Comment (alsora) [inline rclcpp/src/rclcpp/executors/events_queue.cpp:106]: Same as above.
If the executor manages the wait and the mutex, the queue only need to expose `empty()` and `front()` APIs

**Comment by mauropasse** [comment]
Created a new PR based on this comments: https://github.com/irobot-ros/rclcpp/pull/40
